File: Typescript/src/game.ts
import Apple from "./apple.js";
import GUI from "./gui.js";
import { Input } from "./input.js";
import Node from "./node.js";
import { sleep_ms } from "./sleep.js";
import Snake from "./snake.js";
import { hideCursor, showCursor } from "./terminal.js";

export default class Game {
    private score: number;
    private snake: Snake;
    private apple: Apple;
    private gui: GUI;

    constructor(start_x: number, start_y: number) {
        this.score = 0;
        this.snake = new Snake(
            start_x,
            start_y,
            new Node(start_x, start_y + 1, null),
        );
        this.apple = new Apple(this.snake);
        this.gui = new GUI();
    }

    public async start() {
        hideCursor();

        let dx = 1;
        let dy = 0;
        let running = true;

        const inputListener = async () => {
            while (running) {
                const keyInfo = await Input.waitForKey();
                const key = keyInfo.name;

                switch (key) {
                    case "w":
                    case "W":
                    case "up":
                        if (dy !== 1) {
                            dx = 0;
                            dy = -1;
                        }
                        break;

                    case "s":
                    case "S":
                    case "down":
                        if (dy !== -1) {
                            dx = 0;
                            dy = 1;
                        }
                        break;

                    case "a":
                    case "A":
                    case "left":
                        if (dx !== 1) {
                            dx = -1;
                            dy = 0;
                        }
                        break;

                    case "d":
                    case "D":
                    case "right":
                        if (dx !== -1) {
                            dx = 1;
                            dy = 0;
                        }
                        break;

                    case "q":
                    case "Q":
                    case "escape":
                        running = false;
                        break;
                }
            }
        };

        inputListener();

        while (running) {
            let ax = this.snake.head.x + dx;
            let ay = this.snake.head.y + dy;

            let grow = ax === this.apple.x && ay === this.apple.y;

            if (grow) {
                this.apple.eat(this.snake);
                ++this.score;
            }

            if (!this.snake.move(ax, ay, grow)) {
                break;
            }

            this.gui.genGUI(this.snake, this.apple, this.score);

            await sleep_ms(300);
        }

        GUI.gameOver();
        showCursor();
        await Input.waitForExit();
    }
}

File: lua/game.lua
local Snake = require "snake"
local Node  = require "node"
local Apple = require "apple"
local gui   = require "gui"
local terminal = require "terminal"
local input    = require "input"
local sleep    = require "sleep"

local Game = {}

function Game:new(sx, sy)
    local g = {
        score = 0,
        snake = Snake:new(sx, sy, Node:new(sx, sy + 1, nil)),
        apple = nil,
        gui = gui.new()
    }
    setmetatable(g, self)
    self.__index = self

    g.apple = Apple:new(g.snake)

    return g
end

function Game:start()
    terminal.hide_cursor()
    if input.init then
        input.init()
    end

    local dx = 1
    local dy = 0

    while true do
        local key = input.get_key()

        if key == "w" or key == "up" then
            if dy ~= 1 then
                dx = 0
                dy = -1
            end
        end
        if key == "s" or key == "down" then
            if dy ~= -1 then
                dx = 0
                dy = 1
            end
        end
        if key == "a" or key == "left" then
            if dx ~= 1 then
                dx = -1
                dy = 0
            end
        end
        if key == "d" or key == "right" then
            if dx ~= -1 then
                dx = 1
                dy = 0
            end
        end
        if key == "q" then
            break
        end

        local ax = self.snake.head.x + dx
        local ay = self.snake.head.y + dy

        local grow = ax == self.apple.x and ay == self.apple.y
        if grow then
            self.apple:eat(self.snake)
            self.score = self.score + 1
        end

        if not self.snake:move(ax, ay, grow) then
            break
        end

        self.gui:draw(self.snake, self.apple, self.score)

        sleep.ms(300)
    end
    self.gui:game_over()
    input.wait_exit()
    if input.reset then
        input.reset()
    end
    terminal.show_cursor()
end

return Game